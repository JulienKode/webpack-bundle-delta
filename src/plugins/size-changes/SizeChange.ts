/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */
import { BudgetConfig } from './config';

export interface SizeChange {
  file: string;
  prevSize?: number;
  prevGzSize?: number;
  prevBrSize?: number;
  currentSize?: number;
  currentGzSize?: number;
  currentBrSize?: number;
  sizeDiff?: number;
  gzSizeDiff?: number;
  brSizeDiff?: number;
  significantFields: SignificantField[];
  budget?: BudgetResult;
}

export enum BudgetResultType {
  'size' = 'size',
  'gzSize' = 'gzSize',
  'brSize' = 'brSize',
}

export interface BudgetResult {
  amount: number;
  type: BudgetResultType;
  matched: BudgetConfig;
}

export enum SignificantField {
  sizeDiffBytes = 'sizeDiffBytes',
  sizeDiffPercent = 'sizeDiffPercent',
  gzSizeDiffBytes = 'gzSizeDiffBytes',
  gzSizeDiffPercent = 'gzSizeDiffPercent',
  brSizeDiffBytes = 'brSizeDiffBytes',
  brSizeDiffPercent = 'brSizeDiffPercent',
}
