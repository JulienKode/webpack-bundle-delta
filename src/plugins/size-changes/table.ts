/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */
import table from 'markdown-table';
import bytes from 'bytes';
import chalk from 'chalk';
import stringWidth from 'string-width';
import { SizeChange, BudgetResultType, BudgetResult, SignificantField } from './SizeChange';

export enum TableType {
  markdown,
  html,
}

interface BoldFunction {
  (text: string): string;
}

const boldHtml = (text: string) => `<strong>${text}</strong>`;

const formatChangePercentage = (
  diff: number | null,
  significant = false,
  boldFunction: BoldFunction = chalk.bold
) => {
  if (diff === null) {
    return '-';
  }
  if (diff === 0) {
    return '0%';
  }

  const fixedValue = parseFloat(diff.toFixed(2));
  let result = `${fixedValue > 0 ? '+' : ''}${fixedValue}%`;
  if (significant) {
    result = boldFunction(result);
  }
  return result;
};

const formatBytes = (
  value: number | null,
  includeSign = false,
  significant = false,
  boldFunction: BoldFunction = chalk.bold
) => {
  if (value === null) {
    return '-';
  }
  if (value === 0) {
    return '0B';
  }

  let result = `${includeSign && value > 0 ? '+' : ''}${bytes(value, { decimalPlaces: 2 })}`;
  if (significant) {
    result = boldFunction(result);
  }
  return result;
};

const statSizeInfo = (
  prevSize: number,
  currentSize: number,
  sizeDiff: number,
  isNew: boolean,
  isOld: boolean,
  significantFields: SignificantField[],
  boldFunction: BoldFunction
) => {
  if (isNew) {
    return ' (ADDED)';
  }
  if (isOld) {
    return ' (REMOVED)';
  }
  if (currentSize === prevSize) {
    return '';
  }
  return ` (${formatBytes(
    currentSize - prevSize,
    true,
    significantFields.includes(SignificantField.sizeDiffBytes),
    boldFunction
  )} / ${formatChangePercentage(
    sizeDiff,
    significantFields.includes(SignificantField.sizeDiffPercent),
    boldFunction
  )})`;
};

const gzSizeInfo = (
  prevGzSize: number,
  currentGzSize: number,
  gzSizeDiff: number,
  significantFields: SignificantField[],
  boldFunction: BoldFunction
) => {
  if (currentGzSize === prevGzSize) {
    return '';
  }
  return ` (${formatBytes(
    currentGzSize - prevGzSize,
    true,
    significantFields.includes(SignificantField.gzSizeDiffBytes),
    boldFunction
  )} / ${formatChangePercentage(
    gzSizeDiff,
    significantFields.includes(SignificantField.gzSizeDiffPercent),
    boldFunction
  )})`;
};

const brSizeInfo = (
  prevBrSize: number,
  currentBrSize: number,
  brSizeDiff: number,
  significantFields: SignificantField[],
  boldFunction: BoldFunction
) => {
  if (currentBrSize === prevBrSize) {
    return '';
  }
  return ` (${formatBytes(
    currentBrSize - prevBrSize,
    true,
    significantFields.includes(SignificantField.brSizeDiffBytes),
    boldFunction
  )} / ${formatChangePercentage(
    brSizeDiff,
    significantFields.includes(SignificantField.brSizeDiffPercent),
    boldFunction
  )})`;
};

const formatBudget = (budget: BudgetResult) => {
  if (!budget) {
    return '';
  }
  let budgetType = 'Stat';
  if (budget.type === BudgetResultType.brSize) {
    budgetType = 'Brotli';
  } else if (budget.type === BudgetResultType.gzSize) {
    budgetType = 'Gzip';
  }
  return `${+budget.amount.toFixed(2)}% (${budgetType})`;
};

const markdownTable = (data: SizeChange[], includeBudget: boolean) => {
  const headers = ['File', 'Size', 'Gzip size', 'Brotli size'];
  const alignment = ['l', 'c', 'c', 'c'];
  if (includeBudget) {
    headers.push('Budget');
    alignment.push('c');
  }
  return table(
    [
      headers,
      ...data.map(
        ({
          file,
          prevSize,
          prevGzSize,
          prevBrSize,
          currentSize,
          currentGzSize,
          currentBrSize,
          sizeDiff,
          gzSizeDiff,
          brSizeDiff,
          significantFields,
          budget,
        }) => {
          const isNew = currentSize !== null && prevSize === null;
          const isOld = currentSize === null && prevSize !== null;
          const row = [
            file,
            `${formatBytes(currentSize)}${statSizeInfo(
              prevSize,
              currentSize,
              sizeDiff,
              isNew,
              isOld,
              significantFields,
              chalk.bold
            )}`,
            `${formatBytes(currentGzSize)}${
              currentGzSize && !isNew && !isOld
                ? gzSizeInfo(prevGzSize, currentGzSize, gzSizeDiff, significantFields, chalk.bold)
                : ''
            }`,
            `${formatBytes(currentBrSize)}${
              currentBrSize && !isNew && !isOld
                ? brSizeInfo(prevBrSize, currentBrSize, brSizeDiff, significantFields, chalk.bold)
                : ''
            }`,
          ];
          if (includeBudget) {
            row.push(formatBudget(budget));
          }
          return row;
        }
      ),
    ],
    { align: alignment, stringLength: stringWidth }
  );
};

const htmlTable = (data: SizeChange[], includeBudget: boolean): string => {
  return `<table>
  <thead>
    <th>File</th>
    <th>Size</th>
    <th>Gzip size</th>
    <th>Brotli size</th>${
      includeBudget
        ? `
    <th>Budget</th>`
        : ''
    }
  </thead>
  <tbody>
${data
  .map(
    ({
      file,
      prevSize,
      prevGzSize,
      prevBrSize,
      currentSize,
      currentGzSize,
      currentBrSize,
      sizeDiff,
      gzSizeDiff,
      brSizeDiff,
      significantFields,
      budget,
    }) => {
      const isNew = currentSize !== null && prevSize === null;
      const isOld = currentSize === null && prevSize !== null;

      return `    <tr>
      <td>${file}</td>
      <td>${formatBytes(currentSize)}${statSizeInfo(
        prevSize,
        currentSize,
        sizeDiff,
        isNew,
        isOld,
        significantFields,
        boldHtml
      )}</td>
      <td>${formatBytes(currentGzSize)}${
        currentGzSize && !isNew && !isOld
          ? gzSizeInfo(prevGzSize, currentGzSize, gzSizeDiff, significantFields, boldHtml)
          : ''
      }</td>
      <td>${formatBytes(currentBrSize)}${
        currentBrSize && !isNew && !isOld
          ? brSizeInfo(prevBrSize, currentBrSize, brSizeDiff, significantFields, boldHtml)
          : ''
      }</td>${
        includeBudget
          ? `
      <td>${formatBudget(budget)}</td>`
          : ''
      }
    </tr>`;
    }
  )
  .join('\n')}
  </tbody>
</table>`;
};

const displayTable = (data: SizeChange[], tableType: TableType = TableType.markdown): string => {
  const hasBudget = data.some((d) => d.budget);
  const generatedTable =
    tableType === TableType.markdown ? markdownTable(data, hasBudget) : htmlTable(data, hasBudget);
  return generatedTable;
};

export default displayTable;
