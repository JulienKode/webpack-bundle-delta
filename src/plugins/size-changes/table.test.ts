/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */
import table, { TableType } from './table';
import { BudgetResultType, SignificantField } from './SizeChange';

describe('table', () => {
  it('generates markdown table', () => {
    expect(
      table([
        {
          file: 'stationsIndexPage.mjs',
          prevSize: 555927,
          prevGzSize: 0,
          prevBrSize: null,
          currentSize: 556468,
          currentGzSize: 148789,
          currentBrSize: null,
          sizeDiff: 0.09731493523430235,
          gzSizeDiff: 0.15616901256756666,
          brSizeDiff: null,
          significantFields: [SignificantField.sizeDiffBytes, SignificantField.sizeDiffPercent],
        },
      ])
    ).toMatchSnapshot();
  });

  it('generates html table', () => {
    expect(
      table(
        [
          {
            file: 'stationsIndexPage.mjs',
            prevSize: 555927,
            prevGzSize: 0,
            prevBrSize: null,
            currentSize: 556468,
            currentGzSize: 148789,
            currentBrSize: null,
            sizeDiff: 0.09731493523430235,
            gzSizeDiff: 0.15616901256756666,
            brSizeDiff: null,
            significantFields: [SignificantField.brSizeDiffPercent],
          },
        ],
        TableType.html
      )
    ).toMatchSnapshot();
  });

  describe('budget', () => {
    it('generates markdown table', () => {
      expect(
        table([
          {
            file: 'stationsIndexPage.mjs',
            prevSize: 555927,
            prevGzSize: 0,
            prevBrSize: null,
            currentSize: 556468,
            currentGzSize: 148789,
            currentBrSize: null,
            sizeDiff: 0.09731493523430235,
            gzSizeDiff: 0.15616901256756666,
            brSizeDiff: null,
            significantFields: [SignificantField.gzSizeDiffBytes],
            budget: {
              amount: 70,
              matched: {
                chunkName: 'stationsIndexPage.mjs',
              },
              type: BudgetResultType.gzSize,
            },
          },
        ])
      ).toMatchSnapshot();
    });

    it('generates html table', () => {
      expect(
        table(
          [
            {
              file: 'stationsIndexPage.mjs',
              prevSize: 555927,
              prevGzSize: 0,
              prevBrSize: null,
              currentSize: 556468,
              currentGzSize: 148789,
              currentBrSize: null,
              sizeDiff: 0.09731493523430235,
              gzSizeDiff: 0.15616901256756666,
              brSizeDiff: null,
              significantFields: [
                SignificantField.gzSizeDiffPercent,
                SignificantField.sizeDiffPercent,
              ],
              budget: {
                amount: 75.12,
                matched: {
                  chunkName: 'stationsIndexPage.mjs',
                },
                type: BudgetResultType.brSize,
              },
            },
          ],
          TableType.html
        )
      ).toMatchSnapshot();
    });
  });
});
