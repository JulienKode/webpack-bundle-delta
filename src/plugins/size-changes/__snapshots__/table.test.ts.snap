// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table budget generates html table 1`] = `
"<table>
  <thead>
    <th>File</th>
    <th>Size</th>
    <th>Gzip size</th>
    <th>Brotli size</th>
    <th>Budget</th>
  </thead>
  <tbody>
    <tr>
      <td>stationsIndexPage.mjs</td>
      <td>543.43KB (+541B / <strong>+0.1%</strong>)</td>
      <td>145.3KB (+145.3KB / <strong>+0.16%</strong>)</td>
      <td>-</td>
      <td>75.12% (Brotli)</td>
    </tr>
  </tbody>
</table>"
`;

exports[`table budget generates markdown table 1`] = `
"| File                  |           Size           |          Gzip size          | Brotli size |   Budget   |
| :-------------------- | :----------------------: | :-------------------------: | :---------: | :--------: |
| stationsIndexPage.mjs | 543.43KB (+541B / +0.1%) | 145.3KB ([1m+145.3KB[22m / +0.16%) |      -      | 70% (Gzip) |"
`;

exports[`table generates html table 1`] = `
"<table>
  <thead>
    <th>File</th>
    <th>Size</th>
    <th>Gzip size</th>
    <th>Brotli size</th>
  </thead>
  <tbody>
    <tr>
      <td>stationsIndexPage.mjs</td>
      <td>543.43KB (+541B / +0.1%)</td>
      <td>145.3KB (+145.3KB / +0.16%)</td>
      <td>-</td>
    </tr>
  </tbody>
</table>"
`;

exports[`table generates markdown table 1`] = `
"| File                  |           Size           |          Gzip size          | Brotli size |
| :-------------------- | :----------------------: | :-------------------------: | :---------: |
| stationsIndexPage.mjs | 543.43KB ([1m+541B[22m / [1m+0.1%[22m) | 145.3KB (+145.3KB / +0.16%) |      -      |"
`;
